### PR TITLE
PLT-173 Rotate images with a non-standard orientation before generating a thumbnail and preview

### PR DESCRIPTION
**Pull request by hmhealey**
For images uploaded from phones, they often set an orientation tag instead of actually rotating the image. Previews and thumbnails are generated without this tag so they are sideways/upside-down. Instead of setting the tag on the previews (which would cause the dimensions to be incorrect), we now rotate them if necessary before generating stuff.

We don't support mirrored orientations with this (graphics-go doesn't support mirroring for some reason), but at least my phone never produces those so it's probably not something we need yet.

### CODE REVIEW

**Comment by coreyhulen** [comment]
@hmhealey it appears to be failing the gofmt cmd

**Comment by hmhealey** [comment]
Strange that it failed since I think my gofmt changed the line it was complaining about on my end. My gofmt didn't complain when I changed it back though

File: api/file.go [hunk 1]
```diff
@@ -143,25 +146,58 @@ func fireAndForgetHandleImages(filenames []string, fileData [][]byte, teamId, ch
...
-					return
+				width := img.Bounds().Dx()
+				height := img.Bounds().Dy()
+
+				// Get the image's orientation and ignore any errors since not all images will have orientation data
+				orientation, _ := getImageOrientation(fileData[i])
+
+				// Create a temporary image that will be manipulated and then used to make the thumbnail and preview image
+				var temp *image.RGBA
+				if orientation >= 1 && orientation <= 4 {
```
Comment (jwilander) [inline api/file.go]: What do `1` and `4` mean here? Is possible to use constants or an enumerated type to make it more clear?

File: api/file.go [hunk 2]
```diff
@@ -143,25 +146,58 @@ func fireAndForgetHandleImages(filenames []string, fileData [][]byte, teamId, ch
...
+					if err != nil {
+						l4g.Error("Unable to rotate image")
+					}
+				case 6, 7:
+					// rotate 90 degrees CW
+					graphics.Rotate(temp, img, &graphics.RotateOptions{Angle: math.Pi / 2})
+					if err != nil {
+						l4g.Error("Unable to rotate image")
+					}
+				case 1, 2:
```
Comment (jwilander) [inline api/file.go]: Same as comment above, can we use constants for these numbers?

File: api/file.go (hunk 1 above)
Comment (hmhealey) [inline api/file.go]: Yeah I can add an enum for that. I had a comment in there originally with a little diagram for each one as well, but I took that out when I was cleaning up my changes.

**Comment by jwilander** [comment]
+1, nice diagram :)

File: api/file.go [hunk 3]
```diff
@@ -212,6 +270,23 @@ func fireAndForgetHandleImages(filenames []string, fileData [][]byte, teamId, ch
 	}()
 }
 
+func getImageOrientation(imageData []byte) (int, error) {
+	if exifData, err := exif.Decode(bytes.NewReader(imageData)); err != nil {
+		return 1, err
```
Comment (crspeller) [inline api/file.go]: Missed some ones here. These should be Upright correct?

Comment (hmhealey) [inline api/file.go]: Ah that's right. Fixing that shortly

**Comment by crspeller** [comment]
+1

**Pull request merged**